File: src/infrastructure/drivers/onewire.rs
//! DS18B20 temperature sensor via software bitbang OneWire protocol.
//!
//! Pin: GPIO33 (PinDriver::input_output_od with Pull::Up).
//!
//! This module provides:
//! - `OneWireBus` — low-level bitbang protocol (reset, read/write byte,
//!   skip_rom, convert_t, read_scratchpad).
//! - `read_sensor()` — full DS18B20 read sequence returning `Option<f32>`.
//! - `temp_celsius()` — thread-safe access to the last-read value.
//!
//! The temperature read runs in a dedicated `std::thread` (16 KB stack) at
//! 1-second intervals. The main loop reads the atomic via `temp_celsius()`.
//!
//! Timing is via `Ets::delay_us()` — not cycle-accurate but within the
//! DS18B20's forgiving timing tolerances (1–10 µs).

use core::sync::atomic::{AtomicI32, Ordering};

use esp_idf_hal::delay::Ets;
use esp_idf_hal::gpio::{InputOutput, InputPin, OutputPin, PinDriver, Pull};

use log::warn;

use crate::config;

/// Sentinel value indicating "sensor disconnected".
const DISCONNECTED: i32 = i32::MIN;

/// Last valid temperature reading in hundredths of a degree Celsius
/// (e.g., 2350 = 23.50 °C). Initialised to `DISCONNECTED`.
static TEMP_C_X100: AtomicI32 = AtomicI32::new(DISCONNECTED);

/// Return the last valid temperature reading, if any.
///
/// Returns `None` if the sensor has never been read or is disconnected.
#[allow(clippy::cast_precision_loss)]
pub fn temp_celsius() -> Option<f32> {
    let v = TEMP_C_X100.load(Ordering::Relaxed);
    if v == DISCONNECTED {
        None
    } else {
        Some(v as f32 / 100.0)
    }
}

/// Clear the temperature reading (set to disconnected).
pub fn clear() {
    TEMP_C_X100.store(DISCONNECTED, Ordering::Relaxed);
}

/// Software bitbang OneWire bus master.
///
/// Implements the DS18B20 protocol using `Ets::delay_us` for precise timing.
/// Timings follow the Dallas Semiconductor DS18B20 datasheet.
pub struct OneWireBus {
    pin: PinDriver<'static, InputOutput>,
}

impl OneWireBus {
    /// Create a new OneWire bus on the given GPIO pin.
    ///
    /// The pin is configured as open-drain with internal pull-up, then set HIGH.
    ///
    /// # Errors
    ///
    /// Returns `EspError` if the GPIO initialisation fails.
    pub fn new(pin: impl InputPin + OutputPin + 'static) -> Result<Self, esp_idf_sys::EspError> {
        let mut pin = PinDriver::input_output_od(pin, Pull::Up)?;
        pin.set_high()?;
        Ok(Self { pin })
    }

    /// Reset the OneWire bus.
    ///
    /// Returns `true` if a device presence pulse was detected.
    pub fn reset(&mut self) -> bool {
        self.pin.set_low().ok();
        Ets::delay_us(480);
        self.pin.set_high().ok();
        Ets::delay_us(75);
        let present = self.pin.is_low();
        Ets::delay_us(405);
        present
    }

    /// Write a single byte to the OneWire bus.
    pub fn write_byte(&mut self, byte: u8) {
        for i in 0..8 {
            if (byte >> i) & 1 != 0 {
                self.write_bit_1();
            } else {
                self.write_bit_0();
            }
        }
    }

    /// Write a logic 1 bit.
    fn write_bit_1(&mut self) {
        self.pin.set_low().ok();
        Ets::delay_us(6);
        self.pin.set_high().ok();
        Ets::delay_us(64);
    }

    /// Write a logic 0 bit.
    fn write_bit_0(&mut self) {
        self.pin.set_low().ok();
        Ets::delay_us(60);
        self.pin.set_high().ok();
        Ets::delay_us(10);
    }

    /// Read a single byte from the OneWire bus.
    pub fn read_byte(&mut self) -> u8 {
        let mut byte = 0u8;
        for i in 0..8 {
            if self.read_bit() {
                byte |= 1 << i;
            }
        }
        byte
    }

    /// Read a single bit from the OneWire bus.
    fn read_bit(&mut self) -> bool {
        self.pin.set_low().ok();
        Ets::delay_us(3);
        self.pin.set_high().ok();
        Ets::delay_us(10);
        let bit = self.pin.is_high();
        Ets::delay_us(53);
        bit
    }

    /// Send the Skip ROM command (0xCC) — assumes a single device on the bus.
    pub fn skip_rom(&mut self) {
        self.write_byte(0xCC);
    }

    /// Send the Convert T command (0x44) to start a temperature conversion.
    pub fn convert_t(&mut self) {
        self.write_byte(0x44);
    }

    /// Read the 9-byte scratchpad (command 0xBE).
    ///
    /// Returns bytes: temp LSB, temp MSB, TH, TL, config, reserved[3], CRC.
    pub fn read_scratchpad(&mut self) -> [u8; 9] {
        self.write_byte(0xBE);
        let mut buf = [0u8; 9];
        for byte in &mut buf {
            *byte = self.read_byte();
        }
        buf
    }
}

// SAFETY: PinDriver<InputOutput> contains a raw pointer to the GPIO
// peripheral's MMIO register block, which lives at a fixed address.
// Moving this value between threads is safe because the register address
// is identical on all ESP32 cores. No thread-local state is involved.
unsafe impl Send for OneWireBus {}

/// Perform a complete DS18B20 read sequence:
/// reset → skip_rom → convert_t → wait → reset → skip_rom → read_scratchpad.
///
/// On success, stores the temperature in `TEMP_C_X100` and returns `Some(f32)`.
/// On failure (no presence pulse, lost during conversion, out of range),
/// stores `DISCONNECTED` and returns `None`.
///
/// # Blocking
///
/// This function blocks for ~800 ms waiting for the temperature conversion.
/// It MUST be called from a dedicated thread (16 KB stack), not the main loop.
pub fn read_sensor(bus: &mut OneWireBus) -> Option<f32> {
    if !bus.reset() {
        warn!("DS18B20 not detected (no presence pulse)");
        TEMP_C_X100.store(DISCONNECTED, Ordering::Relaxed);
        return None;
    }

    bus.skip_rom();
    bus.convert_t();

    // Wait for conversion to complete (max 750 ms for 12-bit resolution)
    std::thread::sleep(std::time::Duration::from_millis(
        config::TEMP_CONVERSION_WAIT_MS,
    ));

    if !bus.reset() {
        warn!("DS18B20 lost during conversion");
        TEMP_C_X100.store(DISCONNECTED, Ordering::Relaxed);
        return None;
    }

    bus.skip_rom();
    let buf = bus.read_scratchpad();

    let lsb = buf[0];
    let msb = buf[1];
    let temp_raw = u16::from(msb) << 8 | u16::from(lsb);
    let temp = f32::from(temp_raw) / 16.0;

    if !(-55.0..=125.0).contains(&temp) {
        warn!("DS18B20 out of range: {temp}°C");
        TEMP_C_X100.store(DISCONNECTED, Ordering::Relaxed);
        return None;
    }

    #[allow(clippy::cast_possible_truncation)]
    TEMP_C_X100.store((temp * 100.0) as i32, Ordering::Relaxed);
    Some(temp)
}

File: components/infrastructure/src/drivers/onewire.cpp
#include "infrastructure/drivers/onewire.hpp"
#include "esp_log.h"
#include "esp_rom_sys.h"
#include "freertos/FreeRTOS.h"
#include "freertos/task.h"

#include <limits>

static constexpr auto TAG = "onewire";

namespace ecotiter::infrastructure::drivers {

OneWireBus::OneWireBus(gpio_num_t pin)
    : pin_(pin) {
    gpio_set_level(pin_, 1);
}

// CONTRACT: OneWire bitbang timing via esp_rom_delay_us. All delays are
// derived from the DS18B20 datasheet and are tolerant of 1-10 us variance.
// No blocking calls other than the busy-wait delay_us spinloops.

bool OneWireBus::reset() {
    gpio_set_level(pin_, 0);
    esp_rom_delay_us(480);
    gpio_set_level(pin_, 1);
    esp_rom_delay_us(75);
    bool present = gpio_get_level(pin_) == 0;
    esp_rom_delay_us(405);
    return present;
}

void OneWireBus::writeByte(uint8_t byte) {
    for (int i = 0; i < 8; ++i) {
        if ((byte >> i) & 1) {
            writeBit1();
        } else {
            writeBit0();
        }
    }
}

void OneWireBus::writeBit1() {
    gpio_set_level(pin_, 0);
    esp_rom_delay_us(6);
    gpio_set_level(pin_, 1);
    esp_rom_delay_us(64);
}

void OneWireBus::writeBit0() {
    gpio_set_level(pin_, 0);
    esp_rom_delay_us(60);
    gpio_set_level(pin_, 1);
    esp_rom_delay_us(10);
}

uint8_t OneWireBus::readByte() {
    uint8_t byte = 0;
    for (int i = 0; i < 8; ++i) {
        if (readBit()) {
            byte |= static_cast<uint8_t>(1 << i);
        }
    }
    return byte;
}

bool OneWireBus::readBit() {
    gpio_set_level(pin_, 0);
    esp_rom_delay_us(3);
    gpio_set_level(pin_, 1);
    esp_rom_delay_us(10);
    bool bit = gpio_get_level(pin_) != 0;
    esp_rom_delay_us(53);
    return bit;
}

void OneWireBus::skipRom() {
    writeByte(0xCC);
}

void OneWireBus::convertT() {
    writeByte(0x44);
}

std::array<uint8_t, 9> OneWireBus::readScratchpad() {
    writeByte(0xBE);
    std::array<uint8_t, 9> buf{};
    for (auto& b : buf) {
        b = readByte();
    }
    return buf;
}

std::optional<float> readSensor(OneWireBus& bus) { // NOLINT(readability-function-cognitive-complexity) // reason: DS18B20 protocol: reset -> convert -> read scratchpad
    if (!bus.reset()) {
        ESP_LOGD(TAG, "DS18B20 not detected (no presence pulse)");
        gTempCX100.store(std::numeric_limits<int32_t>::min(), std::memory_order_relaxed);
        return std::nullopt;
    }

    bus.skipRom();
    bus.convertT();

    vTaskDelay(pdMS_TO_TICKS(800));

    if (!bus.reset()) {
        ESP_LOGW(TAG, "DS18B20 lost during conversion");
        gTempCX100.store(std::numeric_limits<int32_t>::min(), std::memory_order_relaxed);
        return std::nullopt;
    }

    bus.skipRom();
    auto buf = bus.readScratchpad();

    uint16_t raw = static_cast<uint16_t>(buf[1]) << 8 | buf[0];
    int16_t tempRaw = static_cast<int16_t>(raw);
    float temp = static_cast<float>(tempRaw) / 16.0f;

    if (temp < -55.0f || temp > 125.0f) {
        ESP_LOGW(TAG, "DS18B20 out of range: %.2f C", static_cast<double>(temp));
        gTempCX100.store(std::numeric_limits<int32_t>::min(), std::memory_order_relaxed);
        return std::nullopt;
    }

    gTempCX100.store(static_cast<int32_t>(temp * 100.0f), std::memory_order_relaxed);
    return temp;
}

} // namespace ecotiter::infrastructure::drivers
